### PR TITLE
refactor(codeowners): decouple CODEOWNERS emission from the live tree

### PR DESCRIPTION
**Pull request by dagil-nvidia**
## Summary

Decouples CODEOWNERS emission from the live git tree. Generated ownership is now a pure function of `.github/codeowners/areas.yaml` and `external_contributors.yaml`, so unrelated file additions, deletions, or moves on `main` no longer rewrite CODEOWNERS for an otherwise unchanged policy.

This removes a base-branch race that caused long-running PRs to fail the generated-file drift check.

## Root Cause

The previous generator consulted `git ls-files` while emitting policy:

- minimal-cover optimization changed emitted globs as the tree changed;
- keyword auto-classification and keyword co-ownership walked current paths;
- Dockerfile co-ownership was expanded per file.

That made byte-for-byte CODEOWNERS output depend on repository contents outside the policy change.

## What Changed

- Emit each declared area `path_glob` verbatim and keep `compute_resolution(..., tree=None)` tree-independent.
- Treat file-type rules as stable defaults. They are emitted before nested overrides and shared rules, so explicit routes win under GitHub last-match semantics.
- Preserve YAML declaration order for overlapping file-type rules.
- Remove 41 inert legacy keyword rules and reject future non-empty keyword policy instead of silently ignoring it.
- Keep `build_codeowners.py --strict` as the only tree-aware step, limited to validation and coverage reporting.
- Materialize `lib/sglang-sidecar/` as an explicit SGLang path and regenerate CODEOWNERS.
- Keep the workflow drift check policy-only while retaining strict current-tree coverage validation.
- Add regression coverage for tree independence, precedence, declaration order, malformed policy, and deterministic output.

## Routing Impact — Review Requested

Compared with current `origin/main`, effective owner sets change on 33 tracked paths. Coverage remains 4555/4555 explicitly owned paths.

Seven Dockerfiles now retain their explicit nested or shared owners instead of being shadowed by the blanket ops file-type rule. This is the intended precedence model: file-type ownership supplies a default, while explicit policy remains authoritative.

Please review the 33-path routing delta as a policy change even though the generator and coverage gates are green.

## Validation

- [x] `.github/codeowners/test_codeowners.py`: 74 passed on Python 3.12.
- [x] Strict coverage: 4555/4555 tracked paths explicitly owned; no catch-all ownership.
- [x] Generated CODEOWNERS and contributor artifacts are byte-identical after regeneration.
- [x] Black 23.1, Ruff, and full pre-commit passed.
- [x] Live CI passed on DCO-corrected head `c00a2e66b7`.

---

<a href="https://nvidia.devinenterprise.com/review/ai-dynamo/dynamo/pull/11614" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

## Summary by CodeRabbit

- **Chores**
  - Expanded ownership coverage for the SGLang backend paths.
  - Regenerated repository ownership rules with more granular subsystem and path assignments.
  - Updated Dockerfile ownership handling to apply consistently across directory levels.

- **Refactor**
  - Made ownership resolution and generated files deterministic and independent of the current file tree.
  - Improved coverage reporting, including clearer detection of unmatched and unused ownership patterns.

- **Tests**
  - Added validation for deterministic ownership resolution, generation, coverage enforcement, and external contributor assignments.

### CODE REVIEW

**Comment by dagil-nvidia** [comment]
## Co-owner routing delta (coverage unchanged; these are owner-SET changes only)

Regenerating `CODEOWNERS` from the tree-independent emitter drops the old tree-walked keyword/filetype **co-ownership**, so **34 files** change their owner *set*. **Coverage is unchanged**: every one of the **4526** tracked paths still resolves to at least one owner both before and after (uncovered old = 0, uncovered new = 0). No file becomes unowned.

**Method.** Enumerated every tracked path on the PR head tree (`git ls-files`) and resolved each path's owner set against the base `CODEOWNERS` and the regenerated `CODEOWNERS` using GitHub last-match-wins semantics via the repo's own canonical matcher (`.github/codeowners/codeowners_match.py`: `parse_codeowners` + `resolve_owners`). Cross-checked with both the pre- and post-PR versions of the matcher: identical 34-path result set.

### Transitions (owner set: before -> after)

| # files | Before (owner set) | After (owner set) | Team(s) dropped |
|:------:|:-------------------|:------------------|:----------------|
| 11 | `observability`, `runtime` | `runtime` | `observability` |
| 7 | `ops`, `performance` | `ops` | `performance` |
| 6 | `frontend`, `runtime` | `runtime` | `frontend` |
| 2 | `epp`, `ops` | `ops` | `epp` |
| 2 | `fault-tolerance`, `ops` | `ops` | `fault-tolerance` |
| 1 | `ops`, `xpu` | `ops` | `xpu` |
| 1 | `operator`, `ops` | `ops` | `operator` |
| 1 | `operator`, `planner` | `ops` | `operator`, `planner` |
| 1 | `gms`, `ops` | `ops` | `gms` |
| 1 | `ops`, `runtime` | `ops` | `runtime` |
| 1 | `diffusion`, `ops` | `ops` | `diffusion` |

<details><summary>Per-file breakdown (all 34 paths, grouped by transition)</summary>

**`observability`, `runtime` -> `runtime`**  |  11 file(s)  |  drops `observability`

- `lib/runtime/examples/service_metrics/Cargo.toml`
- `lib/runtime/examples/service_metrics/README.md`
- `lib/runtime/examples/service_metrics/src/bin/service_client.rs`
- `lib/runtime/examples/service_metrics/src/bin/service_server.rs`
- `lib/runtime/examples/service_metrics/src/lib.rs`
- `lib/runtime/examples/system_metrics/Cargo.toml`
- `lib/runtime/examples/system_metrics/README.md`
- `lib/runtime/examples/system_metrics/src/bin/system_client.rs`
- `lib/runtime/examples/system_metrics/src/bin/system_server.rs`
- `lib/runtime/examples/system_metrics/src/lib.rs`
- `lib/runtime/examples/system_metrics/tests/integration_test.rs`

**`ops`, `performance` -> `ops`**  |  7 file(s)  |  drops `performance`

- `lib/backend-common/examples/mocker/Dockerfile`
- `lib/backend-common/examples/mocker/Dockerfile.frontend`
- `recipes/deepseek-v4/container/sglang/Dockerfile.dsv4.sglang.b200`
- `recipes/deepseek-v4/container/sglang/Dockerfile.dsv4.sglang.gb200`
- `recipes/glm-5-nvfp4/sglang/disagg/efa/Dockerfile.efa`
- `recipes/kimi-k2.5/tokenspeed/agg/nvidia/Dockerfile`
- `recipes/nemotron-3-nano-omni/Dockerfile`

**`frontend`, `runtime` -> `runtime`**  |  6 file(s)  |  drops `frontend`

- `lib/bindings/python/src/dynamo/logits_processing/__init__.py`
- `lib/bindings/python/src/dynamo/logits_processing/base.py`
- `lib/bindings/python/src/dynamo/logits_processing/examples/__init__.py`
- `lib/bindings/python/src/dynamo/logits_processing/examples/forced_sequence.py`
- `lib/bindings/python/src/dynamo/logits_processing/examples/hello_world.py`
- `lib/bindings/python/src/dynamo/logits_processing/examples/temperature.py`

**`epp`, `ops` -> `ops`**  |  2 file(s)  |  drops `epp`

- `deploy/inference-gateway/epp/Dockerfile`
- `deploy/inference-gateway/ext-proc/Dockerfile`

**`fault-tolerance`, `ops` -> `ops`**  |  2 file(s)  |  drops `fault-tolerance`

- `tests/fault_tolerance/hardware/fault_injection_service/agents/gpu_fault_injector/Dockerfile`
- `tests/fault_tolerance/hardware/fault_injection_service/api_service/Dockerfile`

**`ops`, `xpu` -> `ops`**  |  1 file(s)  |  drops `xpu`

- `container/templates/sglang_xpu_framework.Dockerfile`

**`operator`, `ops` -> `ops`**  |  1 file(s)  |  drops `operator`

- `deploy/operator/Dockerfile`

**`operator`, `planner` -> `ops`**  |  1 file(s)  |  drops `operator`, `planner`

- `deploy/power-agent/Dockerfile`

**`gms`, `ops` -> `ops`**  |  1 file(s)  |  drops `gms`

- `deploy/snapshot/Dockerfile`

**`ops`, `runtime` -> `ops`**  |  1 file(s)  |  drops `runtime`

- `examples/backends/tritonserver/Dockerfile`

**`diffusion`, `ops` -> `ops`**  |  1 file(s)  |  drops `diffusion`

- `examples/diffusers/Dockerfile`

</details>

### Teams losing a co-owner on >=1 path (added as reviewers)

| Team | Files where it was dropped |
|:-----|:--------------------------:|
| `@ai-dynamo/dynamo-observability-codeowners` | 11 |
| `@ai-dynamo/dynamo-performance-codeowners` | 7 |
| `@ai-dynamo/dynamo-frontend-codeowners` | 6 |
| `@ai-dynamo/dynamo-epp-codeowners` | 2 |
| `@ai-dynamo/dynamo-operator-codeowners` | 2 |
| `@ai-dynamo/dynamo-fault-tolerance-codeowners` | 2 |
| `@ai-dynamo/dynamo-xpu-codeowners` | 1 |
| `@ai-dynamo/dynamo-planner-codeowners` | 1 |
| `@ai-dynamo/dynamo-gms-codeowners` | 1 |
| `@ai-dynamo/dynamo-runtime-codeowners` | 1 |
| `@ai-dynamo/dynamo-diffusion-codeowners` | 1 |

Each team above is being **added as a reviewer** on this PR so it can confirm it is fine giving up implicit co-ownership of those paths. `@ai-dynamo/dynamo-ops-codeowners` (already a reviewer) is the team that *absorbs* the dropped Dockerfile co-ownership.

> Note: `@ai-dynamo/dynamo-runtime-codeowners` is the catch-all owner; its single loss is `examples/backends/tritonserver/Dockerfile` now routing to `ops` only. It is added for completeness.

_Analysis only -- no change to the emitter or `CODEOWNERS` in this comment. No files become unowned; this is purely a reduction in the number of co-owners on 34 paths._

File: .github/codeowners/codeowners_match.py [hunk 1]
```diff
@@ -248,227 +262,103 @@ def load_tree(repo: Path) -> list[str]:
...
+    # explicitly in ``shared`` with a path glob.
+    filetype_shared: list[FiletypeShared] = []
+    for rule in filetype_rules:
+        if rule.get("advisory"):
+            continue
+        pattern = rule.get("pattern")
+        coowner = rule.get("coowner")
+        if not pattern or not coowner:
+            continue
+        filetype_shared.append(FiletypeShared(glob=pattern, owners=[coowner]))
```
[the remark this answers is not in the document]
Comment (dagil-nvidia) [reply .github/codeowners/codeowners_match.py:376]: Confirmed and documented as an intentional policy change. Tree-dependent implicit co-ownership is removed; explicit shared rules are now the stable way to retain it. The current update quantifies 33 owner-set changes on current main with 100% coverage. cfc27bd020 also makes file-type rules defaults so explicit nested and shared policy wins.

[the remark this answers is not in the document]
Comment (dagil-nvidia) [reply .github/codeowners/codeowners_match.py:376]: Fixed in cfc27bd020. File-type defaults now render before nested overrides and shared rules. The XPU Dockerfile resolves to XPU, and the regression test covers both override and shared precedence. Seven previously shadowed Dockerfiles now retain their explicit owners.

File: .github/codeowners/codeowners_match.py [hunk 2]
```diff
@@ -248,227 +262,103 @@ def load_tree(repo: Path) -> list[str]:
...
-    Pure function: same spec + same tree -> same model. No I/O beyond the
-    tree list it's given. Both ``build_codeowners.py`` (coverage gate) and
-    ``emit_codeowners.py`` (generator) call this -- the gate and the artifact
-    cannot disagree because there is no second resolver.
-    """
     catch_all = spec.get("meta", {}).get("catch_all", "")
     raw_areas = spec.get("areas", [])
     classify = spec.get("classify", {}) or {}
-    keyword_rules = classify.get("keyword_rules", []) or []
     filetype_rules: list[FiletypeRule] = classify.get("filetype_rules", []) or []
```
[the remark this answers is not in the document]
Comment (dagil-nvidia) [reply .github/codeowners/codeowners_match.py:347]: Intentional and now fail-loud in cfc27bd020. Keyword auto-classification and co-ownership required the live tree; all 41 inert keyword rules were removed and future non-empty keyword_rules are rejected. The current PR update quantifies the remaining owner-set delta; stable co-ownership must be declared under shared.

**Review by saturley-hall — APPROVED** [review]
